### ISSUE TITLE
non-quantity constant_values parameter to pad

### ISSUE THREAD

**Issue by keewis** [hgrecco/pint#992]
Passing non-quantity parameters to `np.pad` in `constant` mode results in behaviour different from passing `dimensionless` quantities:
```python
In [1]: import numpy as np 
   ...: import pint 
   ...: ureg = pint.UnitRegistry()

In [2]: a = np.arange(5) * ureg.m

In [3]: np.pad(a, (2, 4), mode="constant", constant_values=4)                                                                                                                                                                   
Out[3]: array([4, 4, 0, 1, 2, 3, 4, 4, 4, 4, 4]) <Unit('meter')>

In [4]: np.pad(a, (2, 4), mode="constant", constant_values=4 * ureg.dimensionless)                                                                                                                                              
DimensionalityError: Cannot convert from 'dimensionless' (dimensionless) to 'meter' ([length])
```
I'd expect `[3]` and `[4]` to be equivalent, but one of the checks in `test_pad` seems to expect the current behaviour so I'm not sure if this is a misconception on my part.

**Comment by hgrecco** [comment]
My feeling is that the second behaviour is the correct. Just like this:
```python
>>> import pint
>>> ureg = pint.UnitRegistry()
>>> q = [1,2,3] *ureg.meter
>>> q
<Quantity([1 2 3], 'meter')>
>>> q[1]
<Quantity(2, 'meter')>
>>> q[1] = 3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/grecco/anaconda3/envs/sci/lib/python3.6/site-packages/pint/quantity.py", line 1306, in __setitem__
    '`obj.magnitude[%s] = %s`' % (key, value))
pint.errors.DimensionalityError: Cannot convert from '3' to 'meter'. Assign a quantity with the same dimensionality or access the magnitude directly as `obj.magnitude[1] = 3`
>>> q[1] = 3 * ureg.meter
```

**Comment by keewis** [comment]
hmm... I'd change https://github.com/hgrecco/pint/blob/59d9d86f9aeef5cc30adcda58ab81fd85361eb88/pint/numpy_func.py#L654-L660
to
```python
        def _recursive_convert(arg, unit): 
            if iterable(arg): 
                return tuple(_recursive_convert(a, unit=unit) for a in arg)

            if not _is_quantity(arg):
                arg = arg * unit._REGISTRY.dimensionless
                
            return arg.m_as(unit)
```
which makes the test with `constant_values=(4, 600 * ureg.cm)` raise `DimensionalityError` as expected, but I don't know the reasoning for the current behaviour. Any comments on this, @jthielen?

**Comment by jthielen** [comment]
@keewis I unfortunately do not remember my exact reasoning...all I can piece together are the following:
- using the function padder option requires non-quantities that are treated as if they are in the array's units (rather than dimensionless), so having the `constant_values` and `end_values` behave similar could have been desired
- NumPy does not dispatch on the optional arguments of `pad`, only `array`

But, based on what you've shown, I think this was wrong and that the current behavior is indeed unexpected for `constant_values`/`end_values`, especially given the general rule of thumb that all non-upcast-type non-quantities (except zero) are treated exactly like they are dimensionless. So, this should be corrected! Unfortunately though, your proposed fix doesn't quite work for two reasons:
- it fails with non-multiplicative units
- it does not handle the special "non-quantity zero" (which isn't fixed to have dimensionless units, see #121 / #122).

Would you want to fix those and put in a PR, or would you want me to do so?
